File: 0368-largest-divisible-subset/0368-largest-divisible-subset.cpp
class Solution {
public:
    vector<int> largestDivisibleSubset(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<int> res;
        int l = nums.size();
        if (l == 0) return res;
        vector<int> dp(l, 1); 
        int max_index = 0; 
        int max_size = 1; 
        for (int i = 1; i < l; i++) {
            for (int j = 0; j < i; j++) {
                if (nums[i] % nums[j] == 0) {
                    dp[i] = max(dp[i], dp[j] + 1);
                    if (dp[i] > max_size) {
                        max_size = dp[i];
                        max_index = i;
                    }
                }
            }
        }
        int curr_size = max_size; 
        int curr_num = nums[max_index]; 
        for (int i = max_index; i >= 0; i--) {
            if (curr_size == dp[i] && curr_num % nums[i] == 0) {
                res.push_back(nums[i]);
                curr_size--;
                curr_num = nums[i];
            }
        }
        return res;
    }
};

File: 0368-largest-divisible-subset/0368-largest-divisible-subset.java
class Solution {
    //Learnt from meetcode and editorial
    public List<Integer> largestDivisibleSubset(int[] nums) {
        int len = nums.length;

        if(len == 0){
            return new ArrayList<>();
        }

        List<List<Integer>> divisibleStack = new ArrayList<>();

        for(int i=0; i<len; i++){
            divisibleStack.add(new ArrayList<>());
        }

        Arrays.sort(nums);
        int n = len;

        for(int i = n-1; i>=0; i--){
            List<Integer> tempStack = new ArrayList<>();
            for(int j = i+1; j<n; j++){
                if((nums[j] % nums[i] == 0) && (divisibleStack.get(j).size() > tempStack.size())){
                    tempStack = divisibleStack.get(j);
                }
            }
            divisibleStack.get(i).addAll(tempStack);
            divisibleStack.get(i).add(nums[i]);
        }

        List<Integer> largestSubset = new ArrayList<>();

        for(List<Integer> t : divisibleStack){
            if(t.size() > largestSubset.size()){
                largestSubset = t;
            }
        }

        return largestSubset;
    }
}